src/basicfuns.jl

```diff
@@ -152,9 +152,28 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
+
+Note: different than Maechler (2012), also uses bounds specific to Float32 and Float16.
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(x::Real)
+    threshold = _log1pexp_threshold(x)
+    if x > threshold # log1pexp(x) ≈ x for large x
+        return oftype(threshold, x)
+    else
+    	return log1p(exp(x))
+  	end
+end
+
+# returns a threshold such that log1pexp(x) ≈ x for x > threshold
+@generated function _log1pexp_threshold(::T) where {T<:Real}
```

Comment:
LogExpFunctions is supposed to be a very very lightweight package. Hence I am against adding generated functions if they can be avoided. Again, there are no such issues if you limit this PR to Float64/32/16.

Reply:
Mmm I guess something equivalent can be writen with `@eval` instead of `@generated`.

Reply:
FWIW, I don't see the problem with `@generated`. @devmotion, can you please elaborate?